### PR TITLE
[#3916] feat(catalog-lakehouse-paimon): Support alter table for Paimon Catalog

### PR DESCRIPTION
**Pull request by caican00**
### What changes were proposed in this pull request?

Support alter table for Paimon Catalog.

### Why are the changes needed?

Fix: https://github.com/apache/gravitino/issues/3916

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
New UTs and ITs.

### CODE REVIEW

**Comment by caican00** [comment]
CC @FANNG1

File: catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/PaimonCatalogOperations.java [hunk 1]
```diff
@@ -457,4 +482,52 @@ private NameIdentifier buildPaimonNameIdentifier(NameIdentifier identifier) {
...
+      NameIdentifier identifier, TableChange.RenameTable renameTable)
+      throws NoSuchTableException, IllegalArgumentException {
+    try {
+      paimonCatalogOps.renameTable(
+          identifier.toString(),
+          NameIdentifier.of(identifier.namespace(), renameTable.getNewName()).toString());
+    } catch (Catalog.TableNotExistException e) {
+      throw new NoSuchTableException(e, NO_SUCH_TABLE_EXCEPTION, identifier);
+    } catch (Catalog.TableAlreadyExistException e) {
+      throw new TableAlreadyExistsException(e, TABLE_ALREADY_EXISTS_EXCEPTION, identifier);
```
Comment (FANNG1) [inline catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/PaimonCatalogOperations.java:505]: use newTable name not original table name?

File: catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/utils/TableOpsUtils.java [hunk 2]
```diff
@@ -18,20 +18,127 @@
...
+        UpdateColumnType updateColumnType = (UpdateColumnType) tableChange;
+        return updateColumnType(
+            fieldName(updateColumnType), toPaimonType(updateColumnType.getNewDataType()));
+      }
+    } else if (tableChange instanceof UpdateComment) {
+      return updateComment(((UpdateComment) tableChange).getNewComment());
+    } else if (tableChange instanceof SetProperty) {
+      SetProperty setProperty = ((SetProperty) tableChange);
+      String setPropertyKey = setProperty.getProperty();
+      checkReservedProperties(setPropertyKey);
```
Comment (FANNG1) [inline catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/utils/TableOpsUtils.java:123]: reserved properties are not allowed to alter, no need to check again

File: catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/utils/TableOpsUtils.java [hunk 3]
```diff
@@ -18,20 +18,127 @@
...
+        AddColumn addColumn = (AddColumn) tableChange;
+        String fieldName = fieldName(addColumn);
+        checkColumnCapability(fieldName, addColumn.getDefaultValue(), addColumn.isAutoIncrement());
+        return addColumn(
+            fieldName,
+            toPaimonType(addColumn.getDataType()).copy(addColumn.isNullable()),
+            addColumn.getComment(),
+            move(fieldName, addColumn.getPosition()));
+      } else if (tableChange instanceof DeleteColumn) {
+        return dropColumn(fieldName((DeleteColumn) tableChange));
```
Comment (FANNG1) [inline catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/utils/TableOpsUtils.java:97]: fieldName -> getfieldName ?

File: catalogs/catalog-lakehouse-paimon/src/test/java/org/apache/gravitino/catalog/lakehouse/paimon/utils/TestTableOpsUtils.java [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+          assertEquals("col_1", addColumn.fieldName());
+          assertEquals(DataTypeRoot.INTEGER, addColumn.dataType().getTypeRoot());
+          assertEquals(AddColumn.class.getSimpleName(), addColumn.description());
+          assertNotNull(addColumn.move());
+          assertEquals(MoveType.FIRST, addColumn.move().type());
+          assertEquals("col_1", addColumn.move().fieldName());
+          assertNull(addColumn.move().referenceFieldName());
+          assertFalse(addColumn.dataType().isNullable());
+        });
+    assertTableChange(
```
Comment (FANNG1) [inline catalogs/catalog-lakehouse-paimon/src/test/java/org/apache/gravitino/catalog/lakehouse/paimon/utils/TestTableOpsUtils.java:100]: could you split the test to make it more clear?

File: catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/utils/TableOpsUtils.java [hunk 5]
```diff
@@ -46,4 +153,39 @@ private static void checkColumnAutoIncrement(String fieldName, boolean autoIncre
...
+  private static String fieldName(ColumnChange columnChange) {
+    return fieldName(columnChange.fieldName());
+  }
+
+  private static Move move(String fieldName, ColumnPosition columnPosition)
+      throws UnsupportedOperationException {
+    if (columnPosition instanceof After) {
+      return Move.after(fieldName, ((After) columnPosition).getColumn());
+    } else if (columnPosition instanceof Default) {
+      return null;
```
Comment (FANNG1) [inline catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/utils/TableOpsUtils.java:182]: does this acttly works? seems `testUpdateColumnPosition` not cover this case.

Comment (caican00) [reply catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/utils/TableOpsUtils.java:182]: 1. for `AddColumn` in Paimon: https://github.com/apache/paimon/blob/master/paimon-core/src/main/java/org/apache/paimon/schema/SchemaManager.java#L245-L254
if the Move is null, then the new column will be added  to the end.

2. `Default` position is not supported for `UpdateColumnPosition` in Paimon

File: catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/PaimonCatalogOperations.java (hunk 1 above)
Comment (caican00) [reply catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/PaimonCatalogOperations.java]: fixed

File: catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/utils/TableOpsUtils.java (hunk 2 above)
Comment (caican00) [reply catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/utils/TableOpsUtils.java]: removed

File: catalogs/catalog-lakehouse-paimon/src/test/java/org/apache/gravitino/catalog/lakehouse/paimon/utils/TestTableOpsUtils.java (hunk 4 above)
Comment (caican00) [reply catalogs/catalog-lakehouse-paimon/src/test/java/org/apache/gravitino/catalog/lakehouse/paimon/utils/TestTableOpsUtils.java:103]: updated.

File: catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/utils/TableOpsUtils.java (hunk 3 above)
Comment (caican00) [reply catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/utils/TableOpsUtils.java]: update

File: catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/utils/TableOpsUtils.java (hunk 5 above)
Comment (caican00) [reply catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/utils/TableOpsUtils.java:182]: > does this acttly works

https://github.com/apache/gravitino/pull/4349/files#diff-6bea46f4cb01e05fd11727d1fddd956728bcb163d08f4be08d39ee7eeb729b9fR532-R533
there is a test case about `AddColumn` with `Default` position.

**Comment by caican00** [comment]
image
tested IT locally.

**Comment by caican00** [comment]
comments have been addressed, could you help review again? @FANNG1

File: catalogs/catalog-lakehouse-paimon/build.gradle.kts [hunk 6]
```diff
@@ -138,7 +138,7 @@ tasks.test {
     dependsOn(tasks.jar)
 
     doFirst {
-      environment("GRAVITINO_CI_HIVE_DOCKER_IMAGE", "datastrato/gravitino-ci-hive:0.1.13")
+      environment("GRAVITINO_CI_HIVE_DOCKER_IMAGE", "datastrato/gravitino-ci-hive:0.1.12")
```
Comment (caican00) [inline catalogs/catalog-lakehouse-paimon/build.gradle.kts:141]: will roll it back.

File: catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/PaimonTablePropertiesMetadata.java [hunk 7]
```diff
@@ -35,15 +35,26 @@
...
+  public static final String ROWKIND_FIELD = "rowkind.field";
+  public static final String PRIMARY_KEY = "primary-key";
+  public static final String PARTITION = "partition";
 
   private static final Map<String, PropertyEntry<?>> PROPERTIES_METADATA;
 
   static {
     List<PropertyEntry<?>> propertyEntries =
         ImmutableList.of(
-            stringReservedPropertyEntry(COMMENT, "The table comment", true),
```
Comment (FANNG1) [inline catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/PaimonTablePropertiesMetadata.java:45]: why remove COMMENT and CREATOR?

File: catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/utils/TableOpsUtils.java (hunk 5 above)
Comment (FANNG1) [reply catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/utils/TableOpsUtils.java:180]: Iceberg did some logic when updateColumnPosition is `default`, but this seems ridiculous after another thought, tracked in #4412

**Comment by FANNG1** [comment]
LGTM except minor comments

File: catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/PaimonTablePropertiesMetadata.java (hunk 7 above)
Comment (caican00) [reply catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/PaimonTablePropertiesMetadata.java:45]: the `COMMENT` property can be updated, i removed it from reserved Properties.
in Paimon, the `CREATOR` should be `OWNER`, i updated it.

**Comment by caican00** [comment]
comments have been fixed cc @FANNG1

File: catalogs/catalog-lakehouse-paimon/build.gradle.kts (hunk 6 above)
Comment (caican00) [reply catalogs/catalog-lakehouse-paimon/build.gradle.kts]: update.

File: catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/utils/TableOpsUtils.java [hunk 8]
```diff
@@ -18,20 +18,125 @@
...
+    for (TableChange tableChange : tableChanges) {
+      schemaChanges.add(buildSchemaChange(tableChange));
+    }
+    return schemaChanges;
+  }
+
+  public static SchemaChange buildSchemaChange(TableChange tableChange)
+      throws UnsupportedOperationException {
+    if (tableChange instanceof ColumnChange) {
+
```
[the remark this answers is not in the document]
Comment (caican00) [reply catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/utils/TableOpsUtils.java]: update

File: catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/PaimonTablePropertiesMetadata.java (hunk 7 above)
Comment (FANNG1) [reply catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/PaimonTablePropertiesMetadata.java:45]: COMMENT should be updated with `updateComment` not `setProperties`

File: catalogs/catalog-lakehouse-paimon/src/test/java/org/apache/gravitino/catalog/lakehouse/paimon/TestGravitinoPaimonTable.java [hunk 9]
```diff
@@ -383,8 +383,13 @@ void testTableProperty() {
       ops.initialize(
           initBackendCatalogProperties(), entity.toCatalogInfo(), PAIMON_PROPERTIES_METADATA);
       Map<String, String> map = Maps.newHashMap();
-      map.put(PaimonTablePropertiesMetadata.COMMENT, "test");
```
Comment (FANNG1) [inline catalogs/catalog-lakehouse-paimon/src/test/java/org/apache/gravitino/catalog/lakehouse/paimon/TestGravitinoPaimonTable.java:386]: reserve comment here?

File: catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/PaimonTablePropertiesMetadata.java (hunk 7 above)
Comment (caican00) [reply catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/PaimonTablePropertiesMetadata.java:45]: if the key of SetProperty is `COMMENT`,  `updateComment`  works. cc @FANNG1

File: catalogs/catalog-lakehouse-paimon/src/test/java/org/apache/gravitino/catalog/lakehouse/paimon/TestGravitinoPaimonTable.java (hunk 9 above)
Comment (caican00) [reply catalogs/catalog-lakehouse-paimon/src/test/java/org/apache/gravitino/catalog/lakehouse/paimon/TestGravitinoPaimonTable.java:386]: ok, updated.

**Comment by caican00** [comment]
@FANNG1 could you help review again? Thanks!

File: catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/PaimonTablePropertiesMetadata.java (hunk 7 above)
Comment (FANNG1) [reply catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/PaimonTablePropertiesMetadata.java:45]: I think we shouldn't allow user pass 'comment' properties as Gravitino provides `updateComment` interface

Comment (caican00) [reply catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/PaimonTablePropertiesMetadata.java:45]: ok

File: catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/utils/TableOpsUtils.java [hunk 10]
```diff
@@ -18,20 +18,124 @@
...
+        UpdateColumnType updateColumnType = (UpdateColumnType) tableChange;
+        return updateColumnType(
+            getfieldName(updateColumnType), toPaimonType(updateColumnType.getNewDataType()));
+      }
+    } else if (tableChange instanceof UpdateComment) {
+      return updateComment(((UpdateComment) tableChange).getNewComment());
+    } else if (tableChange instanceof SetProperty) {
+      SetProperty setProperty = ((SetProperty) tableChange);
+      String setPropertyKey = setProperty.getProperty();
+      if (COMMENT.equals(setPropertyKey)) {
```
Comment (FANNG1) [inline catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/utils/TableOpsUtils.java:122]: please remove the `comment` logic in `SetProperty` and `RemoveProperty`

Comment (caican00) [reply catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/utils/TableOpsUtils.java]: done

File: catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/PaimonTablePropertiesMetadata.java [hunk 11]
```diff
@@ -35,15 +35,27 @@
...
```
Comment (FANNG1) [inline catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/PaimonTablePropertiesMetadata.java:45]: better to keep `COMMENT` hidden

Comment (caican00) [reply catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/PaimonTablePropertiesMetadata.java:45]: updated.

**Review by FANNG1 — APPROVED** [review]

**Comment by FANNG1** [comment]
@caican00 , could you related add docs to Paimon?

**Comment by caican00** [comment]
> @caican00 , could you related add docs to Paimon?

@FANNG1 yeah, i am doing it.

**Pull request merged**